src/Bicep.Deploy/DeploymentManager.cs

```diff
@@ -2,54 +2,142 @@
-        {
-            var armClient = new ArmClient(new DefaultAzureCredential());
+        this.armClient = armClient;
+    }
 
-            if (deploymentDefinition.SubscriptionId is { } subscriptionId &&
-                subscriptionId.Equals(Guid.Empty.ToString(), StringComparison.OrdinalIgnoreCase))
-            {
-                var subscription = await armClient.GetDefaultSubscriptionAsync(cancellationToken);
-                deploymentDefinition = deploymentDefinition with { SubscriptionId = subscription.Id };
-            }
+    public async Task<ArmDeploymentResource> CreateOrUpdateAsync(ArmDeploymentDefinition deploymentDefinition, CancellationToken cancellationToken)
+    {
+        try
+        {
+            deploymentDefinition = await HandleEmptySubscriptionIdAsync(deploymentDefinition, cancellationToken);
 
-            var deploymentResource = armClient.GetArmDeploymentResource(deploymentDefinition.Id);
+            var deploymentResource = this.armClient.GetArmDeploymentResource(deploymentDefinition.Id);
             var deploymentContent = new ArmDeploymentContent(deploymentDefinition.Properties);
 
             var operation = await deploymentResource.UpdateAsync(WaitUntil.Completed, deploymentContent, cancellationToken);
 
             return operation.Value;
         }
+        catch (RequestFailedException ex)
+        {
+            throw new DeploymentException(ex.Message, ex);
+        }
+    }
+
+    public async Task<ArmDeploymentResource> CreateOrUpdateAsync(ArmDeploymentDefinition deploymentDefinition, Action<ImmutableSortedSet<ArmDeploymentOperation>>? onOperationsUpdated, CancellationToken cancellationToken)
+    {
+        try
+        {
+            deploymentDefinition = await HandleEmptySubscriptionIdAsync(deploymentDefinition, cancellationToken);
+
+            var deploymentResource = this.armClient.GetArmDeploymentResource(deploymentDefinition.Id);
+            var deploymentContent = new ArmDeploymentContent(deploymentDefinition.Properties);
+
+            var deploymentOperation = await deploymentResource.UpdateAsync(WaitUntil.Started, deploymentContent, cancellationToken);
+
+            var currentOperations = ImmutableSortedSet.Create<ArmDeploymentOperation>(Comparer<ArmDeploymentOperation>.Create((x, y) => 
+            {
+                return string.Compare(x.Properties.TargetResource?.Id, y.Properties.TargetResource?.Id, StringComparison.Ordinal);
+            }));
+
+            while (!deploymentOperation.HasCompleted)
+            {
+                // TODO: Handle nested deployments
+                await deploymentOperation.UpdateStatusAsync(cancellationToken);
+                await foreach (var operation in deploymentResource.GetDeploymentOperationsAsync().WithCancellation(cancellationToken))
+                {
+                    currentOperations = currentOperations.Add(operation);
+                    onOperationsUpdated?.Invoke(currentOperations);
```

Comment:
Should we call the callback after the loop?

Reply:
ah yes, good catch!